read-fonts/src/tables/truetype/bytecode/opcode.rs

```diff
@@ -0,0 +1,842 @@
+    FLIPPT = 0x80,
+    FLIPRGON = 0x81,
+    FLIPRGOFF = 0x82,
+    INS83 = 0x83,
+    INS84 = 0x84,
+    SCANCTRL = 0x85,
+    SDPVTL0 = 0x86,
+    SDPVTL1 = 0x87,
+    GETINFO = 0x88,
+    IDEF = 0x89,
+    ROLL = 0x8A,
+    MAX = 0x8B,
+    MIN = 0x8C,
+    SCANTYPE = 0x8D,
+    INSTCTRL = 0x8E,
+    GETVARIATION = 0x8F,
+    GETDATA = 0x90,
+    INS91 = 0x91,
+    INS92 = 0x92,
+    INS93 = 0x93,
+    INS94 = 0x94,
+    INS95 = 0x95,
+    INS96 = 0x96,
+    INS97 = 0x97,
+    INS98 = 0x98,
+    INS99 = 0x99,
+    INS9A = 0x9A,
+    INS9B = 0x9B,
+    INS9C = 0x9C,
+    INS9D = 0x9D,
+    INS9E = 0x9E,
+    INS9F = 0x9F,
+    INSA0 = 0xA0,
+    INSA1 = 0xA1,
+    INSA2 = 0xA2,
+    INSA3 = 0xA3,
+    INSA4 = 0xA4,
+    INSA5 = 0xA5,
+    INSA6 = 0xA6,
+    INSA7 = 0xA7,
+    INSA8 = 0xA8,
+    INSA9 = 0xA9,
+    INSAA = 0xAA,
+    INSAB = 0xAB,
+    INSAC = 0xAC,
+    INSAD = 0xAD,
+    INSAE = 0xAE,
+    INSAF = 0xAF,
+    PUSHB000 = 0xB0,
+    PUSHB001 = 0xB1,
+    PUSHB010 = 0xB2,
+    PUSHB011 = 0xB3,
+    PUSHB100 = 0xB4,
+    PUSHB101 = 0xB5,
+    PUSHB110 = 0xB6,
+    PUSHB111 = 0xB7,
+    PUSHW000 = 0xB8,
+    PUSHW001 = 0xB9,
+    PUSHW010 = 0xBA,
+    PUSHW011 = 0xBB,
+    PUSHW100 = 0xBC,
+    PUSHW101 = 0xBD,
+    PUSHW110 = 0xBE,
+    PUSHW111 = 0xBF,
+    MDRP00000 = 0xC0,
+    MDRP00001 = 0xC1,
+    MDRP00010 = 0xC2,
+    MDRP00011 = 0xC3,
+    MDRP00100 = 0xC4,
+    MDRP00101 = 0xC5,
+    MDRP00110 = 0xC6,
+    MDRP00111 = 0xC7,
+    MDRP01000 = 0xC8,
+    MDRP01001 = 0xC9,
+    MDRP01010 = 0xCA,
+    MDRP01011 = 0xCB,
+    MDRP01100 = 0xCC,
+    MDRP01101 = 0xCD,
+    MDRP01110 = 0xCE,
+    MDRP01111 = 0xCF,
+    MDRP10000 = 0xD0,
+    MDRP10001 = 0xD1,
+    MDRP10010 = 0xD2,
+    MDRP10011 = 0xD3,
+    MDRP10100 = 0xD4,
+    MDRP10101 = 0xD5,
+    MDRP10110 = 0xD6,
+    MDRP10111 = 0xD7,
+    MDRP11000 = 0xD8,
+    MDRP11001 = 0xD9,
+    MDRP11010 = 0xDA,
+    MDRP11011 = 0xDB,
+    MDRP11100 = 0xDC,
+    MDRP11101 = 0xDD,
+    MDRP11110 = 0xDE,
+    MDRP11111 = 0xDF,
+    MIRP00000 = 0xE0,
+    MIRP00001 = 0xE1,
+    MIRP00010 = 0xE2,
+    MIRP00011 = 0xE3,
+    MIRP00100 = 0xE4,
+    MIRP00101 = 0xE5,
+    MIRP00110 = 0xE6,
+    MIRP00111 = 0xE7,
+    MIRP01000 = 0xE8,
+    MIRP01001 = 0xE9,
+    MIRP01010 = 0xEA,
+    MIRP01011 = 0xEB,
+    MIRP01100 = 0xEC,
+    MIRP01101 = 0xED,
+    MIRP01110 = 0xEE,
+    MIRP01111 = 0xEF,
+    MIRP10000 = 0xF0,
+    MIRP10001 = 0xF1,
+    MIRP10010 = 0xF2,
+    MIRP10011 = 0xF3,
+    MIRP10100 = 0xF4,
+    MIRP10101 = 0xF5,
+    MIRP10110 = 0xF6,
+    MIRP10111 = 0xF7,
+    MIRP11000 = 0xF8,
+    MIRP11001 = 0xF9,
+    MIRP11010 = 0xFA,
+    MIRP11011 = 0xFB,
+    MIRP11100 = 0xFC,
+    MIRP11101 = 0xFD,
+    MIRP11110 = 0xFE,
+    MIRP11111 = 0xFF,
+}
+
+impl Opcode {
+    /// Creates an opcode from the given byte.
+    ///
+    /// There is a 1:1 mapping between bytes and opcodes.
+    #[inline]
+    pub fn from_byte(byte: u8) -> Self {
+        OPCODE_FROM_BYTE[byte as usize]
```

Comment:
given that each possible u8 value is a valid discriminant for our enum, we could simplify this, at the cost of scaring rod, and just write


```rust
unsafe { std:mem::transmute(byte) }
```

I believe we would also need to mark the enum as #[repr(transparent)].

Depending on how smart the compiler is being, this could also end up eliding a bounds check on each call, which could be meaningful.

Alternatively we could also skip that possible bounds check with `unsafe { OPCODE_FROM_BYTE.get_unchecked(&byte) }`.

Reply:
I’m fairly certain llvm will elide the bounds checks because it knows the size of the array and the range of the index value.

That leaves the memory read. The optimizer certainly has enough information to avoid this as well but I’d be surprised if it does. I’ll test this later. Even if not, this array will be hot in L1 so I don’t see it being a perf factor.

I plan to do some benchmarking and profiling on this at some point so we can revisit then.